apps/android/app/src/main/java/chat/simplex/app/views/chat/SendMsgView.kt

```diff
@@ -35,185 +36,56 @@ import androidx.core.view.inputmethod.InputConnectionCompat
-                stopRecordingAndAddAudio()
-                stopRecOnNextClick = false
-              } else {
-                // tapped and didn't hold a finger
-                stopRecOnNextClick = true
-              }
-            },
-            onCancel = startStopRecording,
-            onRelease = startStopRecording
-          )
-          val sendButtonModifier = if (recordingTimeRange.last != 0L)
-            Modifier.clip(CircleShape).background(color)
-          else
-            Modifier
-          IconButton({}, Modifier.size(36.dp), enabled = !cs.inProgress, interactionSource = interactionSource) {
-            Icon(
-              when {
-                recordingTimeRange.last != 0L -> Icons.Outlined.ArrowUpward
-                stopRecOnNextClick -> Icons.Filled.Stop
-                allowedVoiceByPrefs -> Icons.Filled.KeyboardVoice
-                else -> Icons.Outlined.KeyboardVoice
-              },
-              stringResource(R.string.icon_descr_record_voice_message),
-              tint = when {
-                recordingTimeRange.last != 0L -> Color.White
-                stopRecOnNextClick -> MaterialTheme.colors.primary
-                allowedVoiceByPrefs -> MaterialTheme.colors.primary
-                else -> HighOrLowlight
-              },
-              modifier = Modifier
-                .size(36.dp)
-                .padding(4.dp)
-                .then(sendButtonModifier)
-            )
-          }
-          DisposableEffect(Unit) {
-            onDispose {
-              rec.stop()
-            }
-          }
+  Box(Modifier.padding(vertical = 8.dp)) {
+    val cs = composeState.value
+    val attachEnabled = !composeState.value.editing
+    val showProgress = cs.inProgress && (cs.preview is ComposePreview.ImagePreview || cs.preview is ComposePreview.VoicePreview || cs.preview is ComposePreview.FilePreview)
+    val showVoiceButton = cs.message.isEmpty() && showVoiceRecordIcon && attachEnabled &&
+        (cs.preview is ComposePreview.NoPreview || (cs.preview is ComposePreview.VoicePreview && recState.value.isStarted))
+    NativeKeyboard(composeState, textStyle, onMessageChange)
+    // Disable clicks on text field
+    if (cs.preview is ComposePreview.VoicePreview) {
+      Box(Modifier.matchParentSize().clickable(enabled = false, onClick = { }))
+    }
+    Box(Modifier.align(Alignment.BottomEnd)) {
+      val permissionsState = rememberMultiplePermissionsState(listOf(Manifest.permission.RECORD_AUDIO))
+      when {
+        showProgress -> ProgressIndicator()
+        showVoiceButton -> when {
+          needToAllowVoiceToContact ->
+            DisallowedVoiceButton { showNeedToAllowVoiceAlert(allowVoiceToContact) }
+          !allowedVoiceByPrefs ->
+            DisallowedVoiceButton { showDisabledVoiceAlert(isDirectChat) }
+          !permissionsState.allPermissionsGranted ->
+            DisallowedVoiceButton { permissionsState.launchMultiplePermissionRequest() }
```

Comment:
these choices can actually be inside the view callback (or determine which callback is passed, this is not that important, but no need to repeat DisallowedVoiceButton three times here.

Reply:
Made like this but I wouldn't do that because it adds two step check which is not good. We won nothing here with such change, only need to give more attention to two checks instead of to one check.
https://github.com/simplex-chat/simplex-chat/pull/1511/commits/1d56e5b217cd51558ec91741f4888373d78c6aea